IWYU forward declares type used in array which causes compilation failure

If I have the following declaration in a header (C):

```
int sd_dhcp_server_set_lpr(sd_dhcp_server *server, const struct in_addr lpr[], size_t n);
```

Then IWYU chooses to forward declare `struct in_addr;`, which then causes the following compilation failure:

```
../src/systemd/sd-dhcp-server.h:72:47: error: array has incomplete element type 'const struct in_addr'
   72 |                 const struct in_addr addresses[],
      |                                               ^
../src/systemd/sd-dhcp-lease.h:30:8: note: forward declaration of 'struct in_addr'
   30 | struct in_addr;
```

--- a/iwyu.cc
+++ b/iwyu.cc
@@ -2673,6 +2673,16 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     return true;
   }
 
+  bool VisitArrayType(ArrayType* type) {
+    if (CanIgnoreCurrentASTNode())
+      return true;
+    if (!compiler()->getLangOpts().CPlusPlus) {
+      const Type* elem_type = type->getElementType().getTypePtr();
+      ReportTypeUse(CurrentLoc(), elem_type, DerefKind::None);
+    }
+    return true;
+  }
+
   //------------------------------------------------------------
   // Visitors of attributes.
 

--- a/tests/c/array.c
+++ b/tests/c/array.c
@@ -1,0 +1,35 @@
+//===--- array.c - test input file for iwyu -------------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// IWYU_ARGS: -I .
+
+#include "tests/c/direct.h"
+
+// IWYU: Indirect is defined in...*indirect.h
+// IWYU: Indirect needs a declaration
+void array(const struct Indirect array[]);
+
+// IWYU: Indirect is defined in...*indirect.h
+typedef struct Indirect (*ArrayPtr)[5];
+
+// IWYU: Indirect is defined in...*indirect.h
+extern const struct Indirect extern_array[];
+
+/**** IWYU_SUMMARY
+
+tests/c/array.c should add these lines:
+#include "tests/c/indirect.h"
+
+tests/c/array.c should remove these lines:
+- #include "tests/c/direct.h"  // lines XX-XX
+
+The full include-list for tests/c/array.c:
+#include "tests/c/indirect.h"  // for Indirect
+
+***** IWYU_SUMMARY */

--- a/tests/c/direct.h
+++ b/tests/c/direct.h
@@ -1,0 +1,14 @@
+//===--- direct.h - test input file for iwyu ------------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// This file is meant to be directly #included by a .c file.  All it
+// does is #include another file (which the .c file is thus
+// #including indirectly).
+
+#include "tests/c/indirect.h"

--- a/tests/c/indirect.h
+++ b/tests/c/indirect.h
@@ -1,0 +1,17 @@
+//===--- indirect.h - test input file for iwyu ----------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// This file is meant to be #included by "direct.h", which is directly
+// #included by some .c file.  It provides an easy-to-access
+// definition of a symbol that will cause an iwyu violation in a .c
+// file.
+
+struct Indirect {
+    int a;
+};

